packages/synapse-react-client/src/utils/hooks/useUploadFileEntity/useUploadFileEntities.ts

```diff
@@ -0,0 +1,423 @@
+import { instanceOfExternalObjectStoreUploadDestination } from '@sage-bionetworks/synapse-client'
+import { FileEntity } from '@sage-bionetworks/synapse-types'
+import pLimit from 'p-limit'
+import { useCallback, useMemo } from 'react'
+import {
+  ProgressCallback,
+  SYNAPSE_STORAGE_LOCATION_ID,
+} from '../../../synapse-client/SynapseClient'
+import {
+  useCreateEntity,
+  useGetDefaultUploadDestination,
+  useUpdateEntity,
+} from '../../../synapse-queries'
+import { FileUploadArgs } from '../../../synapse-queries/file/FileUploadArgs'
+import { useDirectUploadToS3 } from '../../../synapse-queries/file/useDirectUploadToS3'
+import { useSynapseMultipartUpload } from '../../../synapse-queries/file/useSynapseMultipartUpload'
+import { fixDefaultContentType } from '../../ContentTypeUtils'
+import { useSynapseContext } from '../../context/index'
+import useConfirmItems from '../useConfirmItems'
+import {
+  FilePreparedForUpload,
+  PrepareDirsForUploadReturn,
+  usePrepareFileEntityUpload,
+} from './usePrepareFileEntityUpload'
+import {
+  TrackedUploadProgress,
+  useTrackFileUploads,
+} from './useTrackFileUploads'
+
+type UploadFileStatus =
+  | 'PREPARING'
+  | 'UPLOADING'
+  | 'PAUSED'
+  | 'CANCELED_BY_USER'
+  | 'FAILED'
+  | 'COMPLETE'
+
+type Prompt = {
+  title: string
+  message: string
+  onConfirmAll: () => void
+  onConfirm: () => void
+  onSkip: () => void
+  onCancelAll: () => void
+}
+
+type UploaderState =
+  | 'LOADING'
+  | 'WAITING'
+  | 'PROMPT_USER'
+  | 'UPLOADING'
+  | 'COMPLETE'
+  | 'ERROR'
+
+type FileUploadProgress = {
+  file: File
+  progress: ProgressCallback
+  status: UploadFileStatus
+  cancel: () => void
+  pause: () => void
+  resume: () => void
+  remove: () => void
+  failureReason?: string
+}
+
+type UseUploadFileEntitiesReturn = {
+  state: UploaderState
+  errorMessage?: string
+  isPrecheckingUpload: boolean
+  activePrompts: Prompt[]
+  initiateUpload: (files: File[]) => void
+  activeUploadCount: number
+  uploadProgress: FileUploadProgress[]
+}
+
+// Limit the number of concurrent uploads to avoid overwhelming the browser
+// Note that within one upload operation, multiple parts could be uploaded in parallel
+const limitConcurrentUploads = pLimit(8)
```

Comment:
For large set of files, this should limit the number of concurrently computed MD5s (this slowed down my browser with hundreds of files), and prioritize completing file uploads before starting new ones, which IMO is a better user experience. At the same time, I'm trying to avoid adding an artificial throttle.

This is totally unsophisticated, and we may need to tweak this and/or add other limits later.